stock_inventory/models/stock_inventory.py

```diff
@@ -0,0 +1,473 @@
+            domain = self._get_domain_all_quants(base_domain)
+        elif self.product_selection == "manual":
+            domain = self._get_domain_manual_quants(base_domain)
+        elif self.product_selection == "one":
+            domain = self._get_domain_one_quant(base_domain)
+        elif self.product_selection == "lot":
+            domain = self._get_domain_lot_quants(base_domain)
+        elif self.product_selection == "category":
+            domain = self._get_domain_category_quants(base_domain)
+        return self.env["stock.quant"].search(domain)
+
+    def _get_base_domain(self, locations):
+        return (
+            [
+                ("location_id", "in", locations.mapped("id")),
+            ]
+            if self.exclude_sublocation
+            else [
+                ("location_id", "child_of", locations.child_internal_location_ids.ids),
+            ]
+        )
+
+    def _get_domain_all_quants(self, base_domain):
+        return base_domain
+
+    def _get_domain_manual_quants(self, base_domain):
+        self.ensure_one()
+        return expression.AND(
+            [base_domain, [("product_id", "in", self.product_ids.ids)]]
+        )
+
+    def _get_domain_one_quant(self, base_domain):
+        self.ensure_one()
+        return expression.AND(
+            [
+                base_domain,
+                [
+                    ("product_id", "in", self.product_ids.ids),
+                ],
+            ]
+        )
+
+    def _get_domain_lot_quants(self, base_domain):
+        self.ensure_one()
+        return expression.AND(
+            [
+                base_domain,
+                [
+                    ("product_id", "in", self.product_ids.ids),
+                    ("lot_id", "in", self.lot_ids.ids),
+                ],
+            ]
+        )
+
+    def _get_domain_category_quants(self, base_domain):
+        self.ensure_one()
+        return expression.AND(
+            [
+                base_domain,
+                [
+                    "|",
+                    ("product_id.categ_id", "=", self.category_id.id),
+                    ("product_id.categ_id", "in", self.category_id.child_id.ids),
+                ],
+            ]
+        )
+
+    def refresh_stock_quant_ids(self):
+        for rec in self:
+            rec.stock_quant_ids = rec._get_quants(rec.location_ids)
+
+    def _get_quant_joined_names(self, quants, field):
+        return ", ".join(quants.mapped(f"{field}.display_name"))
+
+    def action_state_to_in_progress(self):
+        self.ensure_one()
+        search_filter = [
+            (
+                "location_id",
+                "child_of" if not self.exclude_sublocation else "in",
+                self.location_ids.ids,
+            ),
+            ("to_do", "=", True),
+        ]
+
+        if self.product_ids:
+            search_filter.append(("product_id", "in", self.product_ids.ids))
+            error_field = "product_id"
+            error_message = _(
```

Comment:
```suggestion
            error_message = self.env._(
```
For getting some performance improvement in some cases. See https://github.com/odoo/odoo/pull/174844.

You can do the same in all occurrences

Reply:
Done